File: src/ch/vat.ts
/**
 * Swiss VAT number.
 *
 * Format: CHE + 9 digits + suffix (MWST|TVA|IVA|TPV).
 * The numeric part is validated as a UID.
 *
 * @see https://www.ch.ch/en/value-added-tax-number-und-business-identification-number/
 */

import { clean } from "#util/clean";
import { err } from "#util/result";

import type { ValidateResult, Validator } from "../types";
import { validate as validateUid } from "./uid";

const SUFFIXES = ["MWST", "TVA", "IVA", "TPV"] as const;

const compact = (value: string): string =>
  clean(value, " -./").toUpperCase();

const extractParts = (
  v: string,
): { uid: string; suffix: string } | undefined => {
  for (const s of SUFFIXES) {
    if (v.endsWith(s)) {
      return { uid: v.slice(0, -s.length), suffix: s };
    }
  }
  return undefined;
};

const validate = (value: string): ValidateResult => {
  const v = compact(value);

  const parts = extractParts(v);
  if (!parts) {
    return err(
      "INVALID_FORMAT",
      "Swiss VAT must end with MWST, TVA, IVA, or TPV",
    );
  }

  const uidResult = validateUid(parts.uid);
  if (!uidResult.valid) {
    return uidResult;
  }

  return { valid: true, compact: v };
};

const format = (value: string): string => {
  const v = compact(value);
  const parts = extractParts(v);
  if (!parts) return v;
  const d = parts.uid.slice(3);
  return `CHE-${d.slice(0, 3)}.${d.slice(3, 6)}.${d.slice(6)} ${parts.suffix}`;
};

/** Swiss VAT Number. */
const vat: Validator = {
  name: "Swiss VAT Number",
  localName: "Mehrwertsteuernummer",
  abbreviation: "MWST",
  country: "CH",
  entityType: "company",
  compact,
  format,
  validate,
};

export default vat;
export { compact, format, validate };

File: crates/stdnum-core/src/validators/ch/vat.rs
//! Swiss VAT Number.

use crate::{
  ValidationResult, Validator, compact_without,
  types::{CountryCode, EntityType, ValidatorScope, ValidatorSpec},
  validators::common::{
    generate_from_examples, invalid_checksum, invalid_format, is_digits,
  },
};

const EXAMPLES: &[&str] = &["CHE107787577IVA"];
const SUFFIXES: &[&str] = &["MWST", "TVA", "IVA", "TPV"];

pub static VALIDATOR: Validator = Validator::new(ValidatorSpec {
  id: "ch.vat",
  name: "Swiss VAT Number",
  local_name: "Mehrwertsteuernummer",
  abbreviation: "MWST",
  aliases: &["MWST", "TVA", "IVA"],
  candidate_pattern: r"CHE-?\d{3}\.?\d{3}\.?\d{3}\s?(?:MWST|TVA|IVA)",
  scope: ValidatorScope::Country(CountryCode::Ch),
  entity_type: EntityType::Company,
  source_url: Some("https://www.estv.admin.ch/"),
  lengths: &[],
  examples: EXAMPLES,
  compact,
  format,
  validate,
  generate: Some(generate),
  parse: None,
});

#[must_use]
pub fn compact(value: &str) -> String {
  compact_without(value.trim(), &[' ', '-', '.', '/', '(', ')']).to_uppercase()
}

#[must_use]
pub fn format(value: &str) -> String {
  let value = compact(value);
  let suffix = SUFFIXES
    .iter()
    .find(|suffix| value.ends_with(**suffix))
    .copied()
    .unwrap_or("");
  let digits = value
    .get(3..value.len().saturating_sub(suffix.len()))
    .unwrap_or("");
  format!(
    "CHE-{}.{}.{} {suffix}",
    digits.get(..3).unwrap_or(""),
    digits.get(3..6).unwrap_or(""),
    digits.get(6..).unwrap_or("")
  )
}

pub fn validate(value: &str) -> ValidationResult {
  let value = compact(value);
  let Some(suffix) = SUFFIXES.iter().find(|suffix| value.ends_with(**suffix))
  else {
    return Err(invalid_format());
  };
  let uid = value
    .get(..value.len().saturating_sub(suffix.len()))
    .unwrap_or("");
  if !uid.starts_with("CHE")
    || uid.len() != 12
    || !uid.get(3..).is_some_and(is_digits)
  {
    return Err(invalid_format());
  }
  if crate::validators::ch::uid::validate(uid).is_err() {
    return Err(invalid_checksum());
  }
  Ok(value)
}

#[must_use]
pub fn generate() -> String {
  generate_from_examples(EXAMPLES, compact, validate)
}
